File: 2021/src/day07/main.py
import os
dir_path = os.path.dirname(os.path.realpath(__file__))


def part1(inp):
    min_ = min(inp)
    max_ = max(inp)
    result = 0
    best_result = 9999999999999

    for i in range(min_, max_ + 1):
        for val in inp:
            result += abs(i - val)

        best_result = min(result, best_result)
        result = 0

    return best_result


def part2(inp):
    min_ = min(inp)
    max_ = max(inp)
    result = 0
    best_result = 9999999999999

    for i in range(min_, max_ + 1):
        for val in inp:
            cost = ((abs(val - i) + 1) * abs(val - i)) // 2
            result += abs(cost)

        best_result = min(result, best_result)
        result = 0

    return best_result


def main():
    with open(f'{dir_path}/../../inputs/day07/input') as f:
        inp = list(map(int, f.read().strip().split(',')))

    print(part1(inp[:]))
    print(part2(inp[:]))


if __name__ == '__main__':
    main()

File: 2021/cpp/day07/main.cpp
#include "../utils.h"

int main()
{
    vector<ll> ps = readInts(cin, ',');
    printCont(ps);

    ll _min = 1e9;
    ll _max = 0;
    for (auto p: ps)
    {
        _min = min(_min, p);
        _max = max(_max, p);
    }

    // Part 1
    ll result = 1e9;
    for (int i = _min; i <= _max; ++i)
    {
        ll cost = 0;
        for (auto& p: ps) cost += abs(i - p); 
        result = min(result, cost);
    }

    print(result);

    // Part 2
    result = 1e9;
    for (int i = _min; i <= _max; ++i)
    {
        ll cost = 0;
        for (auto& p: ps) cost += abs(i - p) * (abs(i - p) + 1) / 2; 
        result = min(result, cost);
    }
    print(result);

    return 0;
}
